### DISCUSSION TITLE
_bcr enhancement for mobile use

### DISCUSSION THREAD

**Discussion by Adam-Veritas** [cortexrd/Knack-Toolkit-Library#408] [Ideas]
I am interested in the QR code reading capability built into KTL.  I have been experimenting with the _bcr keyword and have gotten it to work with a barcode reader connected to a PC.  This works fine, but it is not the solution that I had envisioned for my application.  I use QR codes for inventory management.  The items in inventory each have a unique QR code and the inventory locations each have a unique QR code.  I would like to be able to open a page in Knack on a mobile device and be able to launch the camera app to scan the QR code and then populate the field in Knack with the output of that.  I'd like for there to be an option for fields with the _bcr keyword to display a camera icon next to them that can be clicked to launch the camera app.  The user will then scan the QR code and the output will be passed back to the field on the knack page.  Basically, I'm looking for a way to scan a QR code to enter text into a field rather than manually typing it in.

**Comment by cortexrd** [comment]
Hi @Adam-Veritas,

I've just added support for your use case.  Very cool indeed!

You can try it in **Dev** code for now.
See here if you don't know how to switch to Dev code:  https://github.com/cortexrd/Knack-Toolkit-Library/discussions/186

It introduces a new keyword: **_brcm** for **Barcode Reader Manual**

Actually, it works with both QR and most other code 1D formats.

Usage:

`_bcrm=[field1, field2, etc]`

The fields parameters:

1. are universal selectors, i.e. can be labels of field IDs
2. must be input fields, where you want a scan button added next to them. 
3. none means all fields in form

Works with **ktlRoles** also.

Let me know how it goes,
Norm

**Comment by cortexrd** [comment]
You can try it here in my KTL Tutorials / Demo page:

https://ctrnd.knack.com/ktl-tutorials#ktl-demo-page/barcodes/manual-reading/

Hint: On you desktop browser, right-click in the page, then: **Create QR Code for this page**

![image](https://github.com/user-attachments/assets/031f128d-d8ab-424a-8b3b-c22836091e91)

Then point your mobile device's camera to the QR Code and this will lead you to this page, where you can test the feature.

ND

**Comment by Adam-Veritas** [comment]
@cortexrd, wow this is great.  I've spent some time testing this out, and it works as expected for Field inputs.  Is there a way to apply it to a search field?  I tried adding _bcrm=Search to the Description on a Search View, but that didn't seem to work.  Also, can this be made to auto-submit like the _bcr keyword does?

![image](https://github.com/user-attachments/assets/c799324b-f640-43b3-b9e3-4cc470443144)

**Comment by cortexrd** [comment]
Sure, I can add that.  I'm just very busy right now, so I can't promise when.  Probably in the coming days.

**Comment by Adam-Veritas** [comment]
Awesome, thanks!  I actually went ahead and put together some Javascript that does this for my app.  I'd be happy to share it if you think it would be helpful.

**Reply by cortexrd** [reply]
I'd be curious to see it.  If it's not too big, less and 100 lines maybe, post it here.
But I always re-use the internal KTL utilities like `ktl.views.searchDropdown`, if ever you want to check this out.

The parameters are these:

`searchDropdown: function (srchTxt = '', fieldId = '', matchMode, showPopup = true, viewId = '', pfSaveForm = true) {...`

But you only need the first two: text to find (QR code here) and the field ID.

If you can do this, you might be interested in becoming a KTL collaborator?!  I'd be more than happy to invite you.

Norm
